End-to-End-Tests.md

```diff
@@ -1,228 +1,333 @@
+
+If you use one of options 2-4, you should create a chain of element selectors where the top of the chain uses option 1. Suppose we have a DOM like this:
+
+```text
+
+       Root
+       /  \
+      /   ...
+     /      \
+         Element A: class="protractor-test-elem-a"
+             \
+    ...      ...
+            /  \
+           ...  \
+                 \
+              Element B: id="elem-b"
 ```
-    .all(by.css(...)).first()
-    .all(by.css(...)).last()
-    .all(by.css(...)).get(n)
+
+Then you can select Element B with this selector chain:
+
+```js
+var elemB = element(by.css('.protractor-test-elem-a')).element(by.id('elem-b'));
 ```
-in place of `.element(by.css(...))`. If this does not suffice then you may need to iterate over all the candidates, examining each in turn until you find the right one. An example of how to do so is given by `moveToState` in `core/tests/protractor_utils/editor.js`.
 
-**WARNING: This can lead to flaky and fragile tests, so avoid this if possible. See below for details.**
+Notice that the top of the chain, where we select Element A, uses method 1.
 
-All the protractor code assumes you are working in an Angular webpage. If you need to move to non-Angular context (for example in an iframe) then look at the login function of `users.js` or the `embedding.js` test for examples of how to proceed.
+Sometimes you need to distinguish between several different elements which all look the same to element selectors. You can iterate over all the elements to find the right one. For example, suppose we want to click on the button to open a topic, where the button text is the topic name. We could find the right button like this:
+
+```js
+var buttons = element.all(by.css('.protractor-test-button'));
+...
+var openTopic = async function(topicName) {
+  await waitFor.elementToBeClickable(
+    buttons.first(),
+    'Topic buttons taking too long to become clickable.'
+  );
+  for (i = 0; i < buttons.count(); i++) {
+    var button = buttons.get(i);
+    var buttonText = await action.getText(
+      `Topic button ${i}`, button);
+    if (buttonText === topicName) {
+      await action.click('Topic button ${i}', button);
+      return True;
+    }
+  }
+  return False;
+}
+```
 
-It can happen that a test fails because the webpage has not finished loading functionality by the point at which protractor attempts to use it; failures of this kind can be inconsistent between successive runs of the same test. If one occurs you can try the following:
+It might be tempting to use the `.first()`, `.last()`, and `.get(n)` functions directly when you know what order the elements will come in. However, this makes the tests fragile to changes in the page, and it makes the code hard to read. You should also avoid accessing page elements by index since that's not how most users will find elements. They will be relying on the text identifying those elements, and your test should too.
```

Comment:
```suggestion
It might be tempting to use the `.first()`, `.last()`, and `.get(n)` functions directly when you know what order the elements will come in. However, this makes the tests fragile to changes in the page, and it makes the code hard to read. You should also avoid accessing page elements by index because that's not how most users will find elements. They will be relying on the text identifying those elements, and your test should too.
```